feat: Add sandbox controller configuration, disabled by default.

ponimas opened the pull request:
Added the deployment manifest for the sandbox-controller component.

A new cluster configuration item, sandbox_controller_enabled, has been introduced to control its deployment. The controller is disabled by default, and setting this config item to "true" is required to enable its deployment in the cluster.

A conditional deletion rule was added to ensure the deployment is removed when the configuration item is not set to "true".

ponimas committed:
feat: Add sandbox controller

Added the deployment manifest for the sandbox-controller component.

A new cluster configuration item, sandbox_controller_enabled, has been introduced to control its deployment. The controller is disabled by default, and setting this config item to "true" is required to enable its deployment in the cluster.

A conditional deletion rule was added to ensure the deployment is removed when the configuration item is not set to "true".

linki commented on cluster/manifests/sandbox-controller/deployment.yaml:
```diff
@@ -0,0 +1,42 @@
+# {{ $image := "container-registry.zalando.net/gwproxy/sandbox-controller:main-8" }}
+# {{ $version := index (split $image ":") 1 }}
+
+apiVersion: apps/v1
```
This should be guarded by:

```
{{ if eq .Cluster.ConfigItems.sandbox_controller_enabled "true" }}
```

linki commented on cluster/manifests/sandbox-controller/deployment.yaml:
```diff
@@ -0,0 +1,42 @@
+        # prometheus.io/scrape: "true"
+    spec:
+      priorityClassName: "{{ .Cluster.ConfigItems.system_priority_class }}"
+      serviceAccountName: sandbox-controller
```
This ServiceAccount also needs to be created.

ponimas replied:
~Could you point me to the docs on how to do that?~

linki commented:
What's the purpose of this controller?

ponimas commented:
> What's the purpose of this controller?

It will create routegroups/ingresses to allow testing in production.
[Solution design document](https://docs.google.com/document/d/1ytLltsASWhKRfPucAmw_8IWl5NdhRjHidxZ4A1V03nk/edit?tab=t.0#heading=h.cd1shwdl7itu)

ponimas committed:
chore: Conditionally deploy sandbox controller

Wrap the sandbox controller deployment manifest in a conditional block.
This allows enabling or disabling the deployment based on the cluster's
'sandbox_controller_enabled' configuration item.

ponimas committed:
feat: Add Sandbox CRD

Define the Sandbox custom resource definition.
This CRD is necessary for the sandbox controller to operate and manage sandbox environments.

Signed-off-by: Aleksandr Ponimaskin <aleksandr.ponimaskin@zalando.de>

ponimas committed:
feat: Configure sandbox controller

Add necessary RBAC manifests (ServiceAccount, ClusterRole, and
ClusterRoleBinding) for the sandbox controller. This grants the
controller permissions to manage sandbox, routegroup, and ingress resources.

Wrap the sandbox CRD manifest in the same configuration
flag (sandbox_controller_enabled) as the new RBAC manifests. This
ensures that the CRD and RBAC are only deployed if the sandbox
controller is enabled in the cluster configuration.

Signed-off-by: Aleksandr Ponimaskin <aleksandr.ponimaskin@zalando.de>

szuecs commented on cluster/manifests/sandbox-controller/sandbox_crd.yaml:
```diff
@@ -0,0 +1,38 @@
+{{ if eq .Cluster.ConfigItems.sandbox_controller_enabled "true" }}
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
```
I think we need to create this first before we create the rbac

ponimas committed:
refactor: Order sandbox manifests

Rename the sandbox controller manifest files to include numeric prefixes.

This change ensures Kubernetes applies the resources in the correct order: CRD, RBAC, and then the Deployment, preventing potential dependency issues during installation.

szuecs commented:
:+1:

szuecs commented:
error is unrelated

ponimas commented:
👍

AlexanderYastrebov commented on cluster/manifests/sandbox-controller/30-deployment.yaml:
```diff
@@ -0,0 +1,44 @@
+# {{ $image := "container-registry.zalando.net/gwproxy/sandbox-controller:main-8" }}
+# {{ $version := index (split $image ":") 1 }}
```
You may yaml-comment other template instructions like this elsewhere to make manifests a valid yaml and enable e.g. autoformatting.
